### ISSUE TITLE
Allow reusing the Docker container

### ISSUE THREAD

**Issue by rdiez** [ocrmypdf/OCRmyPDF#909]
The documentation for the "OCRmyPDF Docker image" states:

"Unlike typical Docker containers, in this section the OCRmyPDF Docker container is emphemeral – it runs for one OCR job and terminates, just like a command line program. We are using Docker to deliver an application (as opposed to the more conventional case, where a Docker container runs as a server)."

It is not clear whether this short-lived container mode is the only one allowed if you want to use the ocrmypdf command-line tool.

A Docker image is very easy and convenient way to use OCRmyPDF. However, a container takes long to start. I wanted to batch-process a large number of PDF files, and the container start-up time is an unnecessary last.

Is there a way to keep the container running and execute several "ocrmypdf - - <input.pdf >output.pdf" commands on it?

If that is currently not possible, that would be a great feature.

**Comment by jbarlow83** [comment]
You can run it as a web service too - the documentation mentions this somewhere.
Although I don't think the start up time of a container should be a problem compared to several seconds per page of processing.

**Comment by rdiez** [comment]
As stated above, I want to use the ocrmypdf command-line tool, so the web service does not cut it.

Besides, ocrmypdf can be used to optimise the size of PDF files without adding an OCR text overlay, and it seems to do a better job at it than my manual attempts with Ghostscript. If you have many small PDF documents, and you will not be doing OCR on them, the Docker container start-up time becomes significant.

**Comment by jbarlow83** [comment]
```
$ time docker run --rm -i jbarlow83/ocrmypdf --version
13.3.0.post1+g2be8eeec.d20220203

________________________________________________________
Executed in  927.68 millis    fish           external 
   usr time   72.34 millis   48.93 millis   23.42 millis 
   sys time  166.13 millis  138.04 millis   28.10 millis 

$ time ocrmypdf --version
13.3.0

________________________________________________________
Executed in  315.89 millis    fish           external 
   usr time  225.74 millis  928.00 micros  224.82 millis 
   sys time   47.97 millis  635.00 micros   47.33 millis 

```

Docker has an overhead of ~600 ms per execution on my 3 year old Linux machine. Are you getting a very different result? What OS are you using?

**Comment by rdiez** [comment]
$ sudo time --portability docker run --rm -i jbarlow83/ocrmypdf --version
13.2.0.post13+g3dfde479.d20220126
real 1.18
user 0.02
sys 0.05

My PC is older (AMD FX-4300 Quad-Core) and needs over 1 second. I am running Ubuntu 20.04.3.

I haven't got ocrmypdf installed locally. That was the whole point of the Docker image, that I do not have to worry about thinks like building JBIG2.

But even 600 ms per PDF file is too much. If you have 10,000 documents to process, that would be more than one extra hour of time and electricity for no good reason.

How about processing a few files in parallel? ocrmypdf does not seem to load all cores with my smallish test PDFs, so I wanted to start 2 or 3 instances at the same time. I guess that each container instance will need extra RAM.

I do not know much about Docker. Is it really that hard to reuse a Docker container to run a command-line tool? Would it be possible to run a few in parallel on the same container?

**Comment by jbarlow83** [comment]
A Docker container cannot automatically see the files on the host; it's isolated, unless you explicit map some part of the host filesystem into the Docker. When you do this mapping, the container and host both have different sets of user/group IDs among other things. Getting this all to cooperate is not easy, but you have to do that if you want Docker to work on more than one file at a time. Right now the container is set up with its entrypoint as being ocrmypdf, but if you run the container with `docker run ... --entrypoint /bin/bash` you can "jump" inside the container and work with it from there. For example you could run a bash script from inside the container that repeatedly calls ocrmypdf. So it can be done, but there's no convenient way to package it for end users. That's the main reason I have Docker set up the way it is. 

To do what you want I'd have to set up some kind of client/server interface with a persistent server and client command line utility that issues work to the server, but there's already a web interface that does the same thing. Run the webserver and use `curl` to send it files if you want.

You'll need well more than one hour's work of electricity to set up either of those configurations.

Someone else is working on a Ubuntu snap which will provide a sandboxed image with all dependencies, basically what you're looking for - something with exactly the same interface as the command line utility.
https://github.com/ocrmypdf/OCRmyPDF/issues/914

Bottom line is I have to choose my battles and I can't choose this one given the alternatives available.

>How about processing a few files in parallel? ocrmypdf does not seem to load all cores with my smallish test PDFs, so I wanted to start 2 or 3 instances at the same time. I guess that each container instance will need extra RAM.

ocrmypdf parallelizes over page count so it cannot use all cores on small files. It usually works best to run sqrt(num cores) ocrmypdf workers, each limited to sqrt(num cores) with the `--jobs` parameter. e.g. 2 workers * 2 jobs is probably ideal for a 4 core machine.

**Comment by rdiez** [comment]
I have been digging more into Docker containers. How about the following:

First, start a container which just waits forever with command "sleep infinity".

`sudo docker run --detach --entrypoint="sleep"  jbarlow83/ocrmypdf  infinity`

Then you can run several commands in the same container like this:

`sudo docker exec --interactive  <container ID>  ocrmypdf - - <input.pdf >output.pdf`

OCRmyPDF creates a temporary directory with a random name, so it should be possible to run several concurrent instances in the same container.

**Comment by jbarlow83** [comment]
It seems that you found a novel solution you are happy with. If you think other people might want this too I'd be happy to review and accept a PR explaining this in the documentation as an option for advanced users. Otherwise, is this matter resolved?

**Comment by rdiez** [comment]
I am not actually using OCRmyPDF seriously, I am just playing with it a little. I experimented with "docker exec" because I thought it would be a nice thing to learn for Docker containers in general. So I haven't tested the OCRmyPDF container under stress with parallel invocations.

But from your comments I gather that you believe my approach should work, and that you see value in this trick. Otherwise, you wouldn't have mentioned adding it to the documentation.

I also understand that you want to close this issue. It if were my project, I wouldn't close it until the trick is documented for all my users. But I personally do not need anything else in this respect, so I do not mind.

**Comment by jbarlow83** [comment]
I understand the issue is important to you. The question is how important it is to all users. I think, since I've suggested using Docker the way I have for several years, and you're the first to raise this concern... respectfully, I haven't found other users who share this concern. I know there is an alternative now, in addition to the others I outlined above, and will provide documentation if it becomes a more common request. I can't keep up with every request anymore.
